api/monitor/monitorclient/client.go

```diff
@@ -0,0 +1,217 @@
+// TickStatus calls f periodically if the StatusTicker has been set
+func (m *MonitorClient) TickStatus(f func()) {
+	if m.statusTicker == nil {
+		return
+	}
+	for range m.statusTicker.C {
+		f()
+	}
+}
+
+// Stream interface allows sending anything with Send()
+type Stream interface {
+	Send(msg interface{})
+	Open()
+	Close()
+	Reset()
+}
+
+type componentClient struct {
+	log log.Interface
+	mu  sync.RWMutex
+
+	status   []*sendbuffer.Stream
+	uplink   []*sendbuffer.Stream
+	downlink []*sendbuffer.Stream
+
+	setup  func()
+	cancel context.CancelFunc
+}
+
+func (c *componentClient) Open() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.cancel != nil {
+		return
+	}
+	c.setup()
+}
+
+func (c *componentClient) Close() {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	if c.cancel == nil {
+		return
+	}
+	c.cancel()
+}
+
+func (c *componentClient) Reset() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.cancel == nil {
+		return
+	}
+	c.cancel()
+	c.setup()
+}
+
+func (c *componentClient) Send(msg interface{}) {
+	switch msg := msg.(type) {
+	case *gateway.Status, *router.Status, *broker.Status, *networkserver.Status, *handler.Status:
+		if len(c.status) == 0 {
+			return
+		}
+		for _, cli := range c.status {
+			cli.SendMsg(msg)
+		}
+		c.log.Debug("Forwarded status to monitor")
+	case *router.UplinkMessage, *broker.DeduplicatedUplinkMessage:
+		if len(c.uplink) == 0 {
+			return
+		}
+		for _, cli := range c.uplink {
+			cli.SendMsg(msg)
+		}
+		c.log.Debug("Forwarded uplink to monitor")
+	case *router.DeviceActivationRequest:
+		if len(c.uplink) == 0 {
+			return
+		}
+		asUplink := &router.UplinkMessage{
+			Payload:          msg.Payload,
+			Message:          msg.Message,
+			ProtocolMetadata: msg.ProtocolMetadata,
+			GatewayMetadata:  msg.GatewayMetadata,
+			Trace:            msg.Trace,
+		}
+		for _, cli := range c.uplink {
+			cli.SendMsg(asUplink)
+		}
+		c.log.Debug("Forwarded activation as uplink to monitor")
+	case *broker.DeduplicatedDeviceActivationRequest:
+		if len(c.uplink) == 0 {
+			return
+		}
+		asUplink := &broker.DeduplicatedUplinkMessage{
+			Payload:          msg.Payload,
+			Message:          msg.Message,
+			DevEui:           msg.DevEui,
+			AppEui:           msg.AppEui,
+			AppId:            msg.AppId,
+			DevId:            msg.DevId,
+			ProtocolMetadata: msg.ProtocolMetadata,
+			GatewayMetadata:  msg.GatewayMetadata,
+			ServerTime:       msg.ServerTime,
+			Trace:            msg.Trace,
+		}
+		for _, cli := range c.uplink {
+			cli.SendMsg(asUplink)
+		}
+		c.log.Debug("Forwarded activation as uplink to monitor")
+	case *router.DownlinkMessage, *broker.DownlinkMessage:
+		if len(c.uplink) == 0 {
+			return
+		}
+		for _, cli := range c.downlink {
+			cli.SendMsg(msg)
+		}
+		c.log.Debug("Forwarded downlink to monitor")
+	default:
+		c.log.Warnf("Unknown message type: %T", msg)
+	}
+}
+
+func (c *componentClient) run(monitor, stream string, buf *sendbuffer.Stream) {
+	if err := buf.Run(); err != nil && err != context.Canceled {
```

Comment:
https://github.com/TheThingsNetwork/go-utils/pull/33
Requested a review from you ;)